share/vss/vss.go

```diff
@@ -0,0 +1,579 @@
+	*aggregator
+}
+
+// Deal is sent by the Dealer to each participants.
+type Deal struct {
+	SessionID   []byte
+	SecShare    *share.PriShare
+	RndShare    *share.PriShare
+	T           uint32
+	Commitments []abstract.Point
+	Signature   []byte
+}
+
+// Complaint is a message that must be broadcasted to every verifiers when
+// a verifier receives an invalid Deal.
+type Complaint struct {
+	SessionID []byte
+	Index     uint32
+	Deal      *Deal
+	Signature []byte
+}
+
+// Justification is a message that is broadcasted by the Dealer in response to
+// a Complaint. It contains the original Complaint as well as the shares
+// distributed to the complainer.
+type Justification struct {
+	Complaint *Complaint
+	Deal      *Deal
+}
+
+// Approval is a message that is sent if a verifier approves the Deal he
+// received from the Dealer.
+type Approval struct {
+	SessionID []byte
+	Index     uint32
+	Signature []byte
+}
+
+// NewDealer returns a Dealer capable of leading the secret sharing scheme. It
+// does not have to be trusted by other Verifiers. The security parameter t is
+// the number of shares required to reconstruct the secret. It must be between
+// (len(verifiers) + 1)/2 <= t <= len(verifiers).
+func NewDealer(suite abstract.Suite, longterm, secret abstract.Scalar, verifiers []abstract.Point, r cipher.Stream, t int) (*Dealer, error) {
+	d := &Dealer{
+		suite:     suite,
+		long:      longterm,
+		secret:    secret,
+		verifiers: verifiers,
+	}
+	if !validT(t, verifiers) {
+		return nil, fmt.Errorf("dealer: t %d invalid.", t)
+	}
+
+	d.t = t
+
+	H := deriveH(d.suite, d.verifiers)
+	f := share.NewPriPoly(d.suite, d.t, d.secret, r)
+	g := share.NewPriPoly(d.suite, d.t, nil, r)
+	pub := d.suite.Point().Mul(nil, d.long)
+
+	// F = coeff * B
+	F := f.Commit(d.suite.Point().Base())
+	// G = coeff * H
+	G := g.Commit(H)
+
+	C, err := F.Add(G)
+	if err != nil {
+		return nil, err
+	}
+	_, commitments := C.Info()
+
+	d.sessionID, err = sessionID(d.suite, pub, d.verifiers, commitments, d.t)
+	if err != nil {
+		return nil, err
+	}
+
+	d.aggregator = newAggregator(d.suite, pub, d.verifiers, commitments, d.t, d.sessionID)
+	// C = F + G
+	d.deals = make([]*Deal, len(d.verifiers))
+	for i := range d.verifiers {
+		fi := f.Eval(i)
+		gi := g.Eval(i)
+		d.deals[i] = &Deal{
+			SessionID:   d.sessionID,
+			SecShare:    fi,
+			RndShare:    gi,
+			Commitments: commitments,
+			T:           uint32(d.t),
+		}
+		if d.deals[i].Signature, err = sign.Schnorr(d.suite, d.long, msgDeal(d.deals[i])); err != nil {
+			return nil, err
+		}
+	}
+	return d, nil
+}
+
+// Deals returns a list of deal for each verifiers.
+func (d *Dealer) Deals() []*Deal {
+	var out = make([]*Deal, len(d.deals))
+	copy(out, d.deals)
+	return out
+}
+
+// ProcessComplaint analyzes the given complaint. If it's a valid complaint,
+// then it returns a Justification. This response must be broadcasted to every
+// participants. If it's an invalid complaint, it returns an error about the
```

Comment:
> ProcessComplaint checks the given complaint. If the complaint is valid, then a justification is returned which has to be broadcasted to all other participants subsequently. If the complaint is invalid, an error is returned. Verifiers ignore invalid complaints.

As a general recommendation, I would try to stick to present tense if possible and not use future forms (like "will").